[quant][graphmode] Observing input/output values in call site

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #33895 [quant][graphmode] insert prepack/unpack for aten::linear
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops
* #33853 [quant][graphmode] Swap quantized functional linear with aten::linear
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers
* #32816 [quant][graphmode] Add add_relu pattern in skip values
* #32814 [quant][graphmode] Skip quantizing input and output in matched module
* **#33277 [quant][graphmode] Observing input/output values in call site**

Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208787](https://our.internmc.facebook.com/intern/diff/D20208787)

jerryzh168 committed:
[quant][graphmode] Observing input/output values in callsite

Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
This one is still WIP

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[WIP][quant][graphmode] Observing input/output values in callsite"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward meethod of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/intpu of self.conv2.
We'll need to insert observers for input/output values of the graph in callsite instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -545,6 +591,7 @@ bool isWeightOfConvOrLinear(Value* v) {
         v->uses().size() == 1,
         "We only support conv/linear weight being used by one node.");
   }
+  return result;
```
The restriction on only a single node for weight seems too strict. What if I have a model like this:

```python

class sharedWeightModule(nn.Module):
     def __init__(self):
           super(sharedWeightModule, self).__init__()
           self.weight = torch.rand(3,4,5)

      def forward(self, x):
         y = F.conv(x, self.weight)
         z = F.conv(y, self.weight)
         return z
```
will this case be handled correctly?

jerryzh168 replied:
I think this might be OK with current code, just need to write tests for these cases. We can enable it later

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -804,13 +855,142 @@ void InsertObserversHelper::insertObservers(
+  // list of observer modules for output values
+  std::vector<c10::optional<script::Module>> graph_output_observers;
+
+  // if the current graph is the entry point graph(the forward graph
```
How does this work when qconfig_dict is specified only for some modules in a model? If I have:
``` python
class submodule(nn.module):
    def __init__(self):
           super(submodule, self).__init__()
           self.conv = nn.Conv2d(3,4,5)

   def forward(self, x):
          return self.conv(x)

class midModule(nn.module):
    def __init__(self):
           super(midModule, self).__init__()
           self.sub1 =subModule()
           self.sub2 =subModule()
  def forward(self ,x):
        z = self.sub1(x)
        z = F.layernorm(z)
        z = self.sub2(z)
        return z

class testModule(nn.module):
    def __init__(self):
           super(testModule, self).__init__()
           self.mod1 = midModule()
  
   def forward(self, x):
        x = F.relu(x)
        x = self.midModule(x)
        x = F.layernorm(x)
        return x
```
Lets say the qconfig_dict requires us to quantize only self.midModule.sub1 and self.midModule.sub2. Will we modify midModule in this case?
What if the user wants to do eager mode quantization on midModule (excluding subModule1 and 2) later? 
In this case, will we be inserting  activation observers after the call to submodule1 and submodule2 in the main module?
sharedWeightModule

jerryzh168 replied:
if you call quantize_script(testModuleInstance.midModule) then it will be treated as the top level module because we operate on script modules.
If you call quantize_script(testModuleInstance) then the forward of testModuleInstance is the top level module(entry point) and we will not modify midModule

jerryzh168 replied:
I remember we have decided not to support mixing eager mode and graph mode?

raghuramank100 replied:
As long as we can do:
```
quantize_script(testModuleInstance.midModule)
quantize(testModuleInstance) # qconfig for midModule and all its descendants is None
``` 
we are ok. From what you say, this should work.

jerryzh168 replied:
yes we can do this. it will work as long as we don't quantize the same module with both eager and graph

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -695,6 +744,36 @@ void InsertObserversHelper::addIntermediateValuesToSkipObserver(
+    }
+  }
+}
+
```
Suggest refactoring pre-process to have graph cleanup steps (only replaceConv and FuseLinear). The rest of the observer specific steps (addIntermediate etc) can be in a different function.

jerryzh168 replied:
Sure, I'll refactor them later

raghuramank100 replied:
Please file a task for this purpose

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Please see comments, thanks

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -709,21 +757,55 @@ void InsertObserversHelper::addIntermediateValuesToSkipObserver(
+        }
+        for (auto i = 0; i < g->inputs().size(); ++i) {
+          auto* input_val = g->inputs()[i];
+          boundary_value_map_[n->inputs()[i]].push_back(input_val);
```
Nit: `s/inputs()[i]/input(i)/`, same for outputs.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -445,6 +486,11 @@ class InsertObserversHelper {
+  std::unordered_set<Graph*> visited_graph_of_observer_map_;
+  std::unordered_map<Value*, script::Module> observer_for_value_;
+  // Map from values from callsite into the values in the CallMethod graph
+  std::unordered_map<Value*, std::vector<Value*>> boundary_value_map_;
```
It's probably better to use `unordered_set` instead of `vector` here. We never need the values in it in any particular order.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -818,13 +871,142 @@ void InsertObserversHelper::insertObservers(
+        // Need to make sure all values are
+        // configured with same observer
+        if (result) {
+          TORCH_CHECK(
```
Should this be an assertion? Isn't it possible to hit this case with a legal model with a lecal qconfig (i.e. if we have conv in two submodules but requested to quantize it in only one of them)?

jerryzh168 replied:
this is checking qconfig for values, the case would be

```
class M(torch.nn.Module):
  def __init__(self):
       self.conv = torch.nn.Conv2d(...)
       self.fc = torch.nn.Linear(...)
   def forward(self, x):
       x1 = self.conv(x)
       x2 = self.fc(x1)
       return x2
```

And we configure self.conv with QConfig1 and self.fc with QConfig2, so `x1` will have two different qconfigs in this case.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -818,13 +871,142 @@ void InsertObserversHelper::insertObservers(
+      const auto& observer = std::get<1>(observer_attrs);
+      module._ivalue()->setAttr(name, observer.clone_instance()._ivalue());
+    }
+    for (auto& invoked_method : getInvokedMethods(module, method_name)) {
```
Why is this loop under `if (graph_observer_map_.count(graph.get()))`? Shouldn't we always go through the invoked methods?

jerryzh168 replied:
why? `graph_observer_map_` is a cache of modifications we did to the graph if we go through the invoked methods.

In the following code, we are also going through invoked methods, I'll show you below.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -818,13 +871,142 @@ void InsertObserversHelper::insertObservers(
+        auto input_observers = std::get<0>(input_output_observers);
+        auto output_observers = std::get<1>(input_output_observers);
+        for (auto i = 0; i < n->inputs().size(); ++i) {
+          if (input_observers[i] && !graph_inputs_outputs.count(n->inputs()[i]) && !values_to_observe.count(n->inputs()[i])) {
```
Clang-format :-p

jerryzh168 replied:
will setup clang-format in my editor, seems didn't succeed last time..

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -709,21 +757,55 @@ void InsertObserversHelper::addIntermediateValuesToSkipObserver(
+        // add mapping from callsite value to value in called graph
+        for (auto i = 0; i < g->outputs().size(); ++i) {
+          auto* return_val = g->outputs()[i];
+          boundary_value_map_[n->outputs()[i]].push_back(return_val);
```
Shouldn't the graph input/output be used as the key in the map? Same graph can be called from several different places, so the same `return_val` can correspond to different `n->output(i)`, which I assume we want to store in the vector.

jerryzh168 replied:
It's the reverse, this is used when we want to find the qconfig for a Value in caller graph

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -818,13 +871,142 @@ void InsertObserversHelper::insertObservers(
+      if (observer_nodes_.count(n)) {
+        continue;
+      }
+      if (n->kind() == prim::CallMethod) {
```
@ZolotukhinM, here we go through invoked methods.

Yes, we changed the way we deal with invoked methods, basically right now we are following the code flow that's been executed in inlined graph. This is needed because skipping observers for ops like maxpool, avgpool needs context information, that is "whether the input values have been observed or not", I added this support in this PR: https://github.com/pytorch/pytorch/pull/33481

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208787](https://our.internmc.facebook.com/intern/diff/D20208787)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Observing input/output values in call site"


Summary:
Currently we insert observer in the called graph, which is incorrect since graphs can be shared
and the decision of whether to insert observer or not might dependend on where the graph is called.
For example, for a call sequence `self.conv1(self.conv2(x))`, we can't inserting observer correctly
if `self.conv1` and `self.conv2` are sharing the same type in the current implementation, because we insert
observer in the graph of the forward method of Conv2d right now and this call sequence requires us to insert
only one observer for the output of self.conv1/input of self.conv2.
We'll need to insert observers for input/output values of the graph in call site instead.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208787](https://our.internmc.facebook.com/intern/diff/D20208787)

[ghstack-poisoned]

mrshenli commented:
Hey @jerryzh168, looks like this PR breaks py2 and py3 lint. Could you please submit a fix for them?

mrshenli commented:
Fix is coming in #34146
